### PR TITLE
prometheus-shelly-exporter: init at 1.0.0

### PR DESCRIPTION
**Pull request by drupol**
Link to the project: https://github.com/aexel90/shelly_exporter

Tested successfully on my router.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/monitoring/prometheus/shelly-exporter.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  src = fetchFromGitHub {
+    owner = "aexel90";
+    repo = pname;
+    rev = "v" + version;
+    sha256 = "sha256-L0TuBDq5eEahQvzqd1WuvmXuQbbblCM+Nvj15IybnVo=";
+  };
+
+  vendorSha256 = "sha256-BCrge2xLT4b4wpYA+zcsH64a/nfV8+HeZF7L49p2gEw=";
+
+  passthru.tests = { inherit (nixosTests.prometheus-exporters) shelly_exporter; };
```
Comment (LeSuisse) [inline pkgs/servers/monitoring/prometheus/shelly-exporter.nix:16]: The test seems to be missing.

**Review by LeSuisse — CHANGES_REQUESTED** [review]

File: pkgs/servers/monitoring/prometheus/shelly-exporter.nix (hunk 1 above)
Comment (drupol) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix:16]: What should I add then?

**Comment by drupol** [comment]
Setting this PR to `draft` since I just noticed that I forgot to include the NixOS module.

File: pkgs/servers/monitoring/prometheus/shelly-exporter.nix (hunk 1 above)
Comment (LeSuisse) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix:16]: You need to your test in `nixos/tests/prometheus-exporters.nix`.

For our purposes I'm guessing something simple would be good enough e.g. a JSON file containing `{}` for the metrics file and checking the endpoint can be accessed could be good enough. We would need to access the Shelly Cloud to do something more precise.

Comment (drupol) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix:16]: It's not possible to run the `shell_exporter` without a valid JSON file... so it's impossible to do a valid test.

What should we do then?

Comment (LeSuisse) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix:16]: It looks like we can:

```
$ ./result/bin/shelly_exporter -metrics-file=/tmp/test.json
metrics available at http://127.0.0.1:9784/metrics
$ curl http://127.0.0.1:9784/metrics
[...]
$ cat /tmp/test.json
{}
```

File: nixos/modules/services/monitoring/prometheus/exporters/shelly.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      default = "127.0.0.1";
+      example = "127.0.0.1";
+      description = lib.mdDoc ''
+        The address to listen on for HTTP requests
+      '';
+    };
+    listen-port = mkOption {
+      type = types.port;
+      default = 9784;
+      example = 9784;
```
Comment (SuperSandro2000) [inline nixos/modules/services/monitoring/prometheus/exporters/shelly.nix:21]:
```suggestion
```
If example is the same as default we should omit it.

Comment (drupol) [reply nixos/modules/services/monitoring/prometheus/exporters/shelly.nix]: Thanks, just fixed it.

File: pkgs/servers/monitoring/prometheus/shelly-exporter.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{ lib, buildGoModule, fetchFromGitHub, nixosTests }:
+
+buildGoModule rec {
+  pname = "shelly_exporter";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "aexel90";
+    repo = pname;
+    rev = "v" + version;
```
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix]: Thanks, just fixed it.

**Comment by drupol** [comment]
Finally, got the test successful, this PR is ready.

File: pkgs/servers/monitoring/prometheus/shelly-exporter.nix (hunk 1 above)
Comment (drupol) [reply pkgs/servers/monitoring/prometheus/shelly-exporter.nix]: Fixed now.

File: nixos/modules/services/monitoring/prometheus/exporters/shelly.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        Path to the JSON file with the metric definitions
+      '';
+    };
+  };
+  serviceOpts = {
+    serviceConfig = {
+      ExecStart = ''
+        ${pkgs.bash}/bin/bash -c "${pkgs.prometheus-shelly-exporter}/bin/shelly_exporter \
+          -metrics-file ${cfg.metrics-file} \
+          -listen-address ${cfg.listenAddress}:${toString cfg.port}"
```
Comment (LeSuisse) [inline nixos/modules/services/monitoring/prometheus/exporters/shelly.nix:23]:
```suggestion
        ${pkgs.bash}/bin/bash -c "${pkgs.prometheus-shelly-exporter}/bin/shelly_exporter \
          -metrics-file ${cfg.metrics-file} \
          -listen-address ${cfg.listenAddress}:${toString cfg.port}"
```
```suggestion
        ${pkgs.prometheus-shelly-exporter}/bin/shelly_exporter \
          -metrics-file ${cfg.metrics-file} \
          -listen-address ${cfg.listenAddress}:${toString cfg.port}
```

It does not look like the `bash -c` is needed. The service can be started without it.

**Review by LeSuisse — CHANGES_REQUESTED** [review]
The binary and module work.

File: nixos/modules/services/monitoring/prometheus/exporters/shelly.nix (hunk 4 above)
Comment (drupol) [reply nixos/modules/services/monitoring/prometheus/exporters/shelly.nix]: Thanks, I fixed it.

**Review by LeSuisse — APPROVED** [review]
LGTM, waiting OfBorg to merge.

**Pull request merged**